### DISCUSSION TITLE
Use `declare` when defining classes?

### DISCUSSION THREAD

**Discussion by gefjon** [coalton-lang/coalton#618] [Ideas and Proposals]
Corollary to [discussion 616](https://github.com/coalton-lang/coalton/discussions/616).

I'm very unsure whether this is a good idea or not, but it seems like if we want to make this change, as early as possible is the time to do it:

Should typeclass definitions use the `declare` keyword to define method signatures?

e.g. should the `Eq` class be:

```
(define-class (Eq :a)
    "Types which have equality defined."
    (declare == (:a -> :a -> Boolean)))
```

Pros:
- syntactic consistency
- makes grepping for declarations easier

Cons:
- syntactic overhead in class definitions (seems very small to me, but non-zero)
- huuuuuge breaking change

The breakage of this change could be alleviated by allowing both the `declare` version and the bare version; treating both the existing classdef for `Eq` and the one above as equivalent. In my dream, we'd write a version where the `declare` form is preferred, and the bare form signals a style-warning; later we'd upgrade to a warning; and even after that we'd upgrade to a hard error. But like I said, this is just a rough sketch, and I think I could probably be convinced pretty easily that the current syntax is better, or that the cost of switching is too high.

**Comment by ornamentist** [comment]
Excellent idea, and for typeclasses with values/constants as well as functions:

```
(define-class (AddSemigroup :a)
    "An _additive_ _semigroup_ is a set of values equipped with a closed, associative operation called _addition_."

    (declare add (:a -> :a -> :a))
    (add (:a -> :a -> :a)))
```

then

```
(define-class (AddSemigroup :a => AddMonoid :a)
    "An additive monoid is an additive semigroup with an _additive_  _identity_ called ZERO."

    (define zero :a)

    (zero (:a)))
```
